pandas/core/groupby/groupby.py

```diff
@@ -2633,6 +2633,93 @@ def _reindex_output(
 
         return output.reset_index(drop=True)
 
+    def sample(
+        self,
+        n: Optional[int] = None,
+        frac: Optional[float] = None,
+        replace: bool = False,
+        weights=None,
```

Comment:
can you type these

Reply:
the random_state annotation you might want to put in _typing

Reply:
@simonjayhawkins When I try to annotate weights using
```python
weights: Optional[Union[List, Tuple, AnyArrayLike]] = None
```
I get this mypy error
```python
pandas/core/groupby/groupby.py:2700: error: Incompatible types in assignment (expression has type "Series", variable has type "Union[List[Any], Tuple[Any, ...], ExtensionArray, None]")
pandas/core/groupby/groupby.py:2700: error: Incompatible types in assignment (expression has type "Series", variable has type "Union[List[Any], Tuple[Any, ...], Index, None]")
```
for this line of code
```python
weights = Series(weights)
```
But AnyArrayLike includes Series so this should be fine. Am I missing something?
```python
AnyArrayLike = TypeVar("AnyArrayLike", "ExtensionArray", "Index", "Series", np.ndarray)
```

Reply:
I think what you are trying to do would work if this was a Union, but the TypeVar essentially binds the type of `weights` to whatever it is on initial assignment, so would only work in cases where `weights` is a Series throughout

Is there a reason to not just declare weights as a `Sequence` for the end user?

Reply:
@WillAyd Thanks, that makes sense. I think an Optional Sequence + Series is working.